native/src/Surface.cc

```diff
@@ -41,8 +42,128 @@ extern "C" JNIEXPORT jlong JNICALL Java_org_jetbrains_skija_Surface__1nGetCanvas
+    return surface->height();
+}
+
 extern "C" JNIEXPORT jlong JNICALL Java_org_jetbrains_skija_Surface__1nMakeImageSnapshot
   (JNIEnv* env, jclass jclass, jlong ptr) {
     SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
     return reinterpret_cast<jlong>(surface->makeImageSnapshot().release());
 }
+
+extern "C" JNIEXPORT jlong JNICALL Java_org_jetbrains_skija_Surface__1nMakeImageSnapshotR
+  (JNIEnv* env, jclass jclass, jlong ptr, jint left, jint top, jint right, jint bottom) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    return reinterpret_cast<jlong>(surface->makeImageSnapshot({left, top, right, bottom}).release());
+}
+
+extern "C" JNIEXPORT jint JNICALL Java_org_jetbrains_skija_Surface__1nGenerationId
+  (JNIEnv* env, jclass jclass, jlong ptr) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    return surface->generationID();
+}
+
+extern "C" JNIEXPORT jint JNICALL Java_org_jetbrains_skija_Surface__1nReadPixels
+  (JNIEnv* env, jclass jclass, jlong ptr, jlong bitmapPtr, jint srcX, jint srcY) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    SkBitmap* bitmap = reinterpret_cast<SkBitmap*>(static_cast<uintptr_t>(bitmapPtr));
+    return surface->readPixels(*bitmap, srcX, srcY);
+}
+
+extern "C" JNIEXPORT void JNICALL Java_org_jetbrains_skija_Surface__1nWritePixels
+  (JNIEnv* env, jclass jclass, jlong ptr, jlong bitmapPtr, jint x, jint y) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    SkBitmap* bitmap = reinterpret_cast<SkBitmap*>(static_cast<uintptr_t>(bitmapPtr));
+    surface->writePixels(*bitmap, x, y);
+}
+
+extern "C" JNIEXPORT void JNICALL Java_org_jetbrains_skija_Surface__1nFlushAndSubmit
+  (JNIEnv* env, jclass jclass, jlong ptr, jboolean syncCpu) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    surface->flushAndSubmit(syncCpu);
+}
+
+extern "C" JNIEXPORT void JNICALL Java_org_jetbrains_skija_Surface__1nFlush
+  (JNIEnv* env, jclass jclass, jlong ptr) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    surface->flush();
+}
+
+extern "C" JNIEXPORT jboolean JNICALL Java_org_jetbrains_skija_Surface__1nUnique
+  (JNIEnv* env, jclass jclass, jlong ptr) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    return surface->unique();
+}
+
+extern "C" JNIEXPORT void JNICALL Java_org_jetbrains_skija_Surface__1nRef
+  (JNIEnv* env, jclass jclass, jlong ptr) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    surface->ref();
+}
+
+extern "C" JNIEXPORT void JNICALL Java_org_jetbrains_skija_Surface__1nUnref
+  (JNIEnv* env, jclass jclass, jlong ptr) {
+    SkSurface* surface = reinterpret_cast<SkSurface*>(static_cast<uintptr_t>(ptr));
+    surface->unref();
+}
```

Comment:
Those come from SkRefCnt, which is specific for C++ ref counting. I believe there’s no need for it to be avaliable from Java. Do you have an example where it’s needed?

Reply:
Yeah I wasn't sure about that as well. Will remove 👍

Reply:
Done